juju/client/connection.py

```diff
@@ -454,32 +463,44 @@ def connect_params(self):
             # a few times.
             for i in range(0, 4):
                 result = await self.login()
-                macaroonJSON = result.get('discharge-required')
+                response = result['response']
+                macaroonJSON = response.get('discharge-required')
                 if macaroonJSON is None:
-                    self.info = result['response']
+                    self.info = response
                     success = True
                     return result
                 # TODO implement macaroon authentication
-                raise NotImplementedError('macaroon authentication not implemented yet')
+                error_msg = response['discharge-required-error']
```

Comment:
Can we safely assume a response that contains `discharge-required` will always contain `discharge-required-error` as well? I.E. we won't hit a `KeyError` here?

Reply:
That's correct, from what @rogpeppe told me.